projects/igniteui-angular/src/lib/grids/common/pipes.ts

```diff
@@ -355,3 +355,39 @@ export class IgxColumnFormatterPipe implements PipeTransform {
         return formatter(value);
     }
 }
+
+@Pipe({
+    name: 'gridAddRow',
+    pure: true
+})
+export class IgxGridAddRowPipe implements PipeTransform {
+
+    constructor(private gridAPI: GridBaseAPIService<IgxGridBaseDirective & GridType>) { }
+
+    transform(collection: any, parentID: number, enableAddRow: boolean, cancelAddMode: boolean, pipeTrigger: number) {
+        const grid = this.gridAPI.grid;
+        if (!enableAddRow) {
+            return collection;
+        }
+        if (parentID === -1) {
+            const result = collection.map(record => {
+                return grid.isAddRowRecord(record) ? record.recordRef : record;
+            });
+            grid.data = result;
+            return result;
+        }
+        const parentRecordIndex = collection.findIndex(record => record[grid.primaryKey] === parentID);
+        if (cancelAddMode) {
+            collection.splice(parentRecordIndex + 1, 1);
```

Comment:
Directly slicing the collection may result in polluting the original grid data. You should ensure you work with a copy of the data so that the original data remains the same and does not have non-data records after it passes through the pipe.